src/pk_encryption.rs

```diff
@@ -0,0 +1,336 @@
+    /// authenticate the message.
+    pub mac: Vec<u8>,
+    /// The ephemeral [`Curve25519PublicKey`] of the message which was used to
+    /// derive the individual message key.
+    pub ephemeral_key: Curve25519PublicKey,
+}
+
+impl Message {
+    /// Attempt to decode a PkEncryption [`Message`] from a Base64 encoded
+    /// triplet of ciphertext, MAC, and ephemeral key.
+    pub fn from_base64(
+        ciphertext: &str,
+        mac: &str,
+        ephemeral_key: &str,
+    ) -> Result<Self, MessageDecodeError> {
+        Ok(Self {
+            ciphertext: base64_decode(ciphertext)?,
+            mac: base64_decode(mac)?,
+            ephemeral_key: Curve25519PublicKey::from_base64(ephemeral_key)?,
+        })
+    }
+}
+
+/// The decryption component of the PkEncryption support.
+///
+/// This struct allows you to share a public key, enabling others to encrypt
+/// messages that can be decrypted using the corresponding private key.
+pub struct PkDecryption {
+    key: Curve25519SecretKey,
+    public_key: Curve25519PublicKey,
+}
+
+impl PkDecryption {
+    /// Create a new random [`PkDecryption`] object.
+    ///
+    /// This will create a new random [`Curve25519SecretKey`] which is used as
+    /// the long-term
+    pub fn new() -> Self {
+        let key = Curve25519SecretKey::new();
+        let public_key = Curve25519PublicKey::from(&key);
+
+        Self { key, public_key }
+    }
+
+    /// Get the [`Curve25519PublicKey`] which is
+    pub const fn public_key(&self) -> Curve25519PublicKey {
+        self.public_key
+    }
+
+    /// Decrypt a [`Message`] which was encrypted for this [`PkDecryption`]
+    /// object.
+    pub fn decrypt(&self, message: &Message) -> Result<Vec<u8>, Error> {
+        let shared_secret = self.key.diffie_hellman(&message.ephemeral_key);
+
+        let expanded_keys = ExpandedKeys::new_helper(shared_secret.as_bytes(), b"");
+        let cipher_keys = CipherKeys::from_expanded_keys(expanded_keys);
+
+        let hmac = HmacSha256::new_from_slice(cipher_keys.mac_key())
+            .expect("We should be able to create a Hmac object from a 32 byte key");
+
+        // BUG: This is a know issue, we check the MAC of an empty message instead of
+        // updating the `hmac` object with the ciphertext bytes.
+        hmac.verify_truncated_left(&message.mac)?;
+
+        let cipher = Aes256CbcDec::new(cipher_keys.aes_key(), cipher_keys.iv());
+        let decrypted = cipher.decrypt_padded_vec_mut::<Pkcs7>(&message.ciphertext)?;
+
+        Ok(decrypted)
+    }
+
+    /// Create a [`PkDecryption`] object from a slice of bytes.
+    pub fn from_bytes(bytes: &[u8; 32]) -> Self {
+        let key = Curve25519SecretKey::from_slice(bytes);
+        let public_key = Curve25519PublicKey::from(&key);
+
+        Self { key, public_key }
+    }
+
+    /// Export this [`PkDecryption`] object to a slice of bytes.
+    pub fn to_bytes(&self) -> Box<[u8; 32]> {
+        self.key.to_bytes()
+    }
```

Comment:
We may want to avoid dealing with bytes here, and have `from_key` and `key` instead.

It would match what is done for `PkEncryption`.

`impl From<Curve25519SecretKey>` implementation would also be nice.

Reply:
Yeah, you're right, no point in re-exposing the `Curve25519SecretKey` constructors.